politeiad/backendv2/tstorebe/plugins/ticketvote/cmds.go

```diff
@@ -0,0 +1,2976 @@
+// regardless of what the vote bits are. The vote collider and the full cast
+// vote are saved to the backend at the same time. A cast vote is not
+// considered valid unless a corresponding vote collider is present.
+type voteCollider struct {
+	Token  string `json:"token"`  // Record token
+	Ticket string `json:"ticket"` // Ticket hash
+}
+
+func (p *ticketVotePlugin) voteColliderSave(treeID int64, vc voteCollider) error {
+	// Prepare blob
+	be, err := convertBlobEntryFromVoteCollider(vc)
+	if err != nil {
+		return err
+	}
+
+	// Save blob
+	return p.tstore.BlobSave(treeID, *be)
+}
+
+// ballot casts the provided votes concurrently. The vote results are passed
+// back through the results channel to the calling function. This function
+// waits until all provided votes have been cast before returning.
+func (p *ticketVotePlugin) ballot(treeID int64, votes []ticketvote.CastVote, results chan ticketvote.CastVoteReply) {
+	// Cast the votes concurrently
+	var wg sync.WaitGroup
+	for _, v := range votes {
+		// Increment the wait group counter
+		wg.Add(1)
+
+		go func(v ticketvote.CastVote) {
+			// Decrement wait group counter once vote is cast
+			defer wg.Done()
+
+			// Setup cast vote details
+			receipt := p.identity.SignMessage([]byte(v.Signature))
+			cv := ticketvote.CastVoteDetails{
+				Token:     v.Token,
+				Ticket:    v.Ticket,
+				VoteBit:   v.VoteBit,
+				Signature: v.Signature,
+				Receipt:   hex.EncodeToString(receipt[:]),
+			}
+
+			// Declare here to prevent goto errors
+			var (
+				cvr ticketvote.CastVoteReply
+				vc  voteCollider
+			)
+
+			// Save cast vote
+			err := p.castVoteSave(treeID, cv)
+			if err == plugins.ErrDuplicateBlob {
+				// This cast vote has already been saved. Its possible that
+				// a previous attempt to vote with this ticket failed before
+				// the vote collider could be saved. Continue execution so
+				// that we re-attempt to save the vote collider.
+			} else if err != nil {
+				t := time.Now().Unix()
+				log.Errorf("cmdCastBallot: castVoteSave %v: %v", t, err)
+				e := ticketvote.VoteErrorInternalError
+				cvr.Ticket = v.Ticket
+				cvr.ErrorCode = e
+				cvr.ErrorContext = fmt.Sprintf("%v: %v",
+					ticketvote.VoteErrors[e], t)
+				goto sendResult
+			}
+
+			// Save vote collider
+			vc = voteCollider{
+				Token:  v.Token,
+				Ticket: v.Ticket,
+			}
+			err = p.voteColliderSave(treeID, vc)
+			if err != nil {
+				t := time.Now().Unix()
+				log.Errorf("cmdCastBallot: voteColliderSave %v: %v", t, err)
+				e := ticketvote.VoteErrorInternalError
+				cvr.Ticket = v.Ticket
+				cvr.ErrorCode = e
+				cvr.ErrorContext = fmt.Sprintf("%v: %v",
+					ticketvote.VoteErrors[e], t)
+				goto sendResult
+			}
+
+			// Update receipt
+			cvr.Ticket = v.Ticket
+			cvr.Receipt = cv.Receipt
+
+			// Update cast votes cache
+			p.activeVotes.AddCastVote(v.Token, v.Ticket, v.VoteBit)
+
+		sendResult:
+			// Send result back to calling function
+			results <- cvr
```

Comment:
Channels often have unforeseen consequences. Do we need a channel or can we have some sort of shared/locked structure instead?

Or if we do have to use a channel, can't we send a batch in a single channel write?

Reply:
We can probably do a shared structure instead.  I'll give that a go.

Reply:
I am ok doing this post-release but we do need an issue written for it. Channels are evil

Reply:
Ok

Reply:
Issue opened: https://github.com/decred/politeia/issues/1372